pages/design_documentation/cleaner.md

```diff
@@ -5,12 +5,138 @@ toc: false
+hot, i.e. has been just written by the **write-back engine**, it is added to 
+the beginning of the LRU list, tagged with current timestamp. 
+
+If a cache line already exists on the list, it is removed from the list only to 
+be inserted again at the beginning, now with a current timestamp.
+
+After **Wake Up** time has passed, background cleaning thread is woken up and
+tries to perform cleaning. However, cleaning will only happen when cache I/O 
+**activity threshold** time has expired.
+
+If all the conditions are met, a flush list of up to **Flush Max Buffers** 
+is built by removing the cache lines from the back of the LRU list if their 
+**staleness time** has expired.
+
+This list is then provided to the OCF cleaner by issuing *ocf_cleaner_fire()*. 
+OCF cleaner then issues an asynchrounous flushing request, eventually writing 
+the dirty data to the core device.
+
+This process is repeated until there is no further data to clean. The 
+background task is then put to sleep for **Wake Up** time period.
+
+
 ### ACP - Aggressive Cleaning Policy
-### NOP - No Operation (effectively disables background cleaning)
+Aggressive Cleaning Policy takes a rather different approach than ALRU. It organizes
+the data into 100MiB chunks, maintaining a list of them per core device. The 
+chunks are put into "buckets" of varying dirty data percentage. There are 11 
+buckets in the ACP, holding data chunks being 0% dirty,  up to 100% dirty, 
+with a 10% granularity.
+
+
+#### Configuration
+ACP uses these config parameters to tune the behavior of the cleaner.
+
+* **Wake Up**- cleaning thread wake up time. (default: 10s)
+
+* **Flush Max Buffers** - maximum amount of dirty cache lines to flush in one cleaning 
+operation (default: 128 lines)
+
+
+#### Operation
+When the cache is started and core devices are added, ACP creates the chunks lists
+per each core. A chunk is a linear, 100MiB part of the core device address space,
+cache-line addressable. This creates an ACP-specific mapping of cache lines to 
+the chunks. These chunks are then put into a proper buckets, depending on the 
+percentage of dirty cache lines in a chunk.
+
+When a cache line is written and dirty data is created, the dirty lines counter 
+in the chunk the cache line belongs to, is incremented. Since the chunks are placed 
+in the buckets according to the dirty lines percentage in a chunk, a chunk may be
+moved to another bucket once the percentage threshold is met.
+
+After **Wake Up** time has passed, background cleaning thread is woken up and
+tries to perform cleaning. In the ACP, the cleaning logic selects the chunk to 
+clean by scanning backwards across the buckets, i.e. starting from the 100% bucket
+down to the 10% one. 
+As soon as there is a non-empty chunk list within a bucket, first chunk from that 
+list is selected. This chunk will be cleaned only if the core device it belongs 
+to is opened and the next cleaning timestamp is exceeded (a backoff time after 
```

Comment:
"only if the core device it belongs to is opened" unnecessary implementation detail.